File: src/lab_two.rs
use std::ops::Range;

use log::info;
use mpi::traits::{Communicator, CommunicatorCollectives, Root};
use num::Signed;
pub use num_traits::Num;
use rand::distributions::uniform::SampleUniform;

use crate::executor::Executor;

pub trait LabTwo<T>
where
    T: Num
        + std::fmt::Display
        + Signed
        + Copy
        + Default
        + SampleUniform
        + std::clone::Clone
        + std::ops::AddAssign
        + num::FromPrimitive
        + mpi::traits::Equivalence
        + std::cmp::PartialOrd
        + std::ops::MulAssign,
    rand::distributions::Standard: rand::distributions::Distribution<T>,
    f32: From<T>,
{
    fn monte_carlo(
        executor: Executor,
        f: fn(T, T) -> Option<T>,
        x_range: Range<T>,
        y_range: Range<T>,
        n: i32,
    ) -> T {
        use rand::prelude::*;
        let mut rand = rand::thread_rng();
        let mut result = T::from_i32(0).unwrap();
        let t_start = mpi::time();
        let mut i_n = 0;
        let mut s = T::from_i32(0).unwrap();
        for _ in (executor.rank()..n).step_by(executor.size() as usize) {
            i_n += 1;
            if let Some(value) = f(
                rand.gen_range(x_range.clone()),
                rand.gen_range(y_range.clone()),
            ) {
                s += value;
            }
        }
        if executor.rank() != 0 {
            executor
                .communicator()
                .process_at_rank(0)
                .reduce_into(&s, mpi::collective::SystemOperation::sum());
            executor
                .communicator()
                .process_at_rank(0)
                .reduce_into(&i_n, mpi::collective::SystemOperation::sum());
        } else {
            let mut gin = T::from_i32(0).unwrap();
            let mut gsum = T::from_i32(0).unwrap();
            executor.communicator().process_at_rank(0).reduce_into_root(
                &s,
                &mut gsum,
                mpi::collective::SystemOperation::sum(),
            );
            executor.communicator().process_at_rank(0).reduce_into_root(
                &i_n,
                &mut gin,
                mpi::collective::SystemOperation::sum(),
            );
            let v = x_range.end * gin / T::from_i32(n).unwrap();
            result = v * gsum / gin;
            info!("Result: {}, n {}", result, n);
            info!("Time estimated: {}", mpi::time() - t_start);
        }

        result
    }

    fn midpoint_rule(executor: Executor, e: f32, f: fn(T) -> T, bound: (T, T)) -> T {
        let mut result = T::from_i32(0).unwrap();
        let t_start = mpi::time();
        let mut n = 1;
        let mut sq: [T; 2] = [T::from_i32(0).unwrap(), T::from_i32(0).unwrap()];
        let mut delta: f32 = 1.0;
        let mut k = 0;
        while delta > e {
            let points_per_proc = n / executor.size();
            let lb = executor.rank() * points_per_proc;
            let ub = if executor.rank() == executor.size() - 1 {
                n - 1
            } else {
                lb + points_per_proc - 1
            };
            let h = (bound.1 - bound.0) / T::from_i32(n).unwrap();
            let mut s = T::from_i32(0).unwrap();
            for i in lb..=ub {
                s += f(bound.0 + h * T::from_f32(i as f32 + 0.5).unwrap());
            }
            executor.communicator().all_reduce_into(
                &s,
                &mut sq[k as usize],
                mpi::collective::SystemOperation::sum(),
            );
            sq[k as usize] *= h;
            if n > 1 {
                delta = <T as Into<f32>>::into((sq[k as usize] - sq[k as usize ^ 1]).abs()) / 3.0;
            }
            n *= 2;
            k ^= 1;
        }

        if executor.rank() == 0 {
            info!(
                "Result Pi: {}; Runge rule: EPS {}, n {}",
                sq[k] * sq[k],
                e,
                n / 2
            );
            info!("Time estimated: {}\n", mpi::time() - t_start);
            result = sq[k] * sq[k];
        }
        result
    }
}

File: src/lab_two.cpp
#include "lab_two.hpp"

const double e = 0.000001;
double drand(double low, double high) {
  return ((double)rand() * (high - low)) / (double)RAND_MAX + low;
}
void monte_carlo(double (*func)(double, double), double lower_x, double lower_y,
                 double upper_x, double upper_y, int n) {
  int world_size, world_rank;
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  double starttime, endtime;
  starttime = MPI_Wtime();

  srand(world_rank);
  int in = 0;
  double s = 0;
  for (int i = world_rank; i < n; i += world_size) {
    double x = drand(lower_x, upper_x);
    double y = drand(lower_y, upper_y);
    in++;
    double tmp = func(x, y);
    if (tmp == -1) {
      continue;
    }
    s += tmp;
  }
  int gin = 0;
  MPI_Reduce(&in, &gin, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);
  double gsum = 0.0;
  MPI_Reduce(&s, &gsum, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

  if (world_rank == 0) {
    double v = 1;
    if (upper_x != 0 && gin != 0) {
      v = upper_x * gin / n;
    }
    double res = v * gsum / gin;
    printf("Result: %.12f, n %d\n", res, n);
    endtime = MPI_Wtime();
    printf("Time estimated: %f", endtime - starttime);
  }
}

void monte_carlo(double (*func)(double, double), double lower_x, double lower_y,
                 double upper_x, double upper_y, double difference, bool append,
                 BoundDifference bound, int n) {
  int world_size, world_rank;
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  double starttime, endtime;
  starttime = MPI_Wtime();

  srand(world_rank);
  int in = 0;
  double s = 0;
  for (int i = world_rank; i < n; i += world_size) {
    double x;
    double y;
    switch (bound) {
    case BoundDifference::lower_x:
      y = drand(lower_y, upper_y);
      lower_x = append ? y + difference : difference - y;
      x = drand(lower_x, upper_x);
      break;
    case BoundDifference::lower_y:
      x = drand(lower_y, upper_y);
      lower_y = append ? x + difference : difference - x;
      y = drand(lower_x, upper_x);
      break;
    case BoundDifference::upper_x:
      y = drand(lower_y, upper_y);
      upper_x = append ? y + difference : difference - y;
      x = drand(lower_x, upper_x);
      break;
    case BoundDifference::upper_y:
      x = drand(lower_y, upper_y);
      upper_y = append ? x + difference : difference - x;
      y = drand(lower_x, upper_x);
      break;
    case BoundDifference::lower_x_against:
      y = drand(lower_y, upper_y);
      lower_x = append ? y + difference : y - difference;
      x = drand(lower_x, upper_x);
      break;
    case BoundDifference::lower_y_against:
      x = drand(lower_y, upper_y);
      lower_y = append ? x + difference : x - difference;
      y = drand(lower_x, upper_x);
      break;
    case BoundDifference::upper_x_against:
      y = drand(lower_y, upper_y);
      upper_x = append ? y + difference : y - difference;
      x = drand(lower_x, upper_x);
      break;
    case BoundDifference::upper_y_against:
      x = drand(lower_y, upper_y);
      upper_y = append ? x + difference : x - difference;
      y = drand(lower_x, upper_x);
      break;
    };
    in++;
    double tmp = func(x, y);
    if (tmp == -1) {
      continue;
    }
    s += tmp;
  }
  int gin = 0;
  MPI_Reduce(&in, &gin, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);
  double gsum = 0.0;
  MPI_Reduce(&s, &gsum, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

  if (world_rank == 0) {
    double v = 1;
    if (upper_x != 0 && gin != 0) {
      v = upper_x * gin / n;
    }
    double res = v * gsum / gin;
    printf("Result: %.12f, n %d\n", res, n);
    endtime = MPI_Wtime();
    printf("Time estimated: %f", endtime - starttime);
  }
}

void midpoint_rule(double (*func)(double), double a, double b) {
  int world_size, world_rank;
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  double starttime, endtime;
  starttime = MPI_Wtime();
  int n0 = 1;
  int n = n0, k;
  double sq[2], delta = 1;
  for (k = 0; delta > e; n *= 2, k ^= 1) {
    int points_per_proc = n / world_size;
    int lb = world_rank * points_per_proc;
    int ub =
        (world_rank == world_size - 1) ? (n - 1) : (lb + points_per_proc - 1);
    double h = (b - a) / n;
    double s = 0.0;
    for (int i = lb; i <= ub; i++) {
      s += func(a + h * (i + 0.5));
    }
    MPI_Allreduce(&s, &sq[k], 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
    sq[k] *= h;
    if (n > n0) {
      delta = fabs(sq[k] - sq[k ^ 1]) / 3.0;
    }
  }
  if (world_rank == 0) {
    printf("Result Pi: %.12f; Runge rule: EPS %e, n %d\n", sq[k] * sq[k], e,
           n / 2);
    endtime = MPI_Wtime();
    printf("Time estimated: %f\n", endtime - starttime);
  }
}
